.github/workflows/workflow.yaml

```diff
@@ -317,6 +317,60 @@ jobs:
             done
           done
 
+  benchmark:
+    name: "Benchmark"
+    # if: github.event_name == 'pull_request'
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    permissions:
+      contents: write
+      pull-requests: write
+    steps:
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@v3
+
+      - name: Download release benchmark file
+        uses: robinraju/release-downloader@v1.6
+        id: get-latest-benchmark
+        continue-on-error: true
+        with: 
+          latest: true
+          fileName: "release_benchmarks.txt"
+
+      - name: Run benchmarks
+        run: |
+          curl -L -O "https://github.com/kubernetes-sigs/kubebuilder/releases/download/v${KUBEBUILDER_VERSION}/kubebuilder_${KUBEBUILDER_VERSION}_linux_amd64.tar.gz" &&\
+            tar -zxvf kubebuilder_${KUBEBUILDER_VERSION}_linux_amd64.tar.gz &&\
+            sudo mv kubebuilder_${KUBEBUILDER_VERSION}_linux_amd64 /usr/local/kubebuilder
+          make benchmark-test BENCHMARK_FILE_NAME=pr_benchmarks.txt
+        if: ${{ steps.get-latest-benchmark.outcome == 'success' }}
+        env:
+          KUBEBUILDER_VERSION: 2.3.1
+
+      - name: Compare benchmarks
+        id: get-comment-body
+        if: ${{ steps.get-latest-benchmark.outcome == 'success' }}
+        run: |
+          export GOPATH="$HOME/go"
+          PATH="$GOPATH/bin:$PATH"
+          go install golang.org/x/perf/cmd/benchstat@latest
+          benchstat release_benchmarks.txt pr_benchmarks.txt > benchstat.txt
```

Comment:
Do we know if the release benchmark was performed on an identical machine/load profile? If not, how much does that impact the accuracy of the comparison? Are we measuring differences in code performance or machine performance?

Reply:
Yes, Since the release job also runs on `ubuntu-latest` the same as comparing the benchmark job which is a [standard runner in github](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources), we may say that we are trying to compare code performance of update (the PR in question) against what is released in the latest version.

Reply:
> Maybe we could run the benchmark multiple times (separate runs so on different machines) to get a sense of the variance and storing that?

We are doing this with `count=5` 
https://github.com/open-policy-agent/gatekeeper/pull/2432/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R110

Reply:
We need to have some kind of signal that the performance numbers are apples-to-apples, otherwise we'll spend time trying to diagnose a regression that's really just an architecture difference.

I'm open to whatever gives us confidence this is the case. Unfortunately thread count alone probably isn't sufficient, since not all threads are created equal.

TBH, my preferred order (of alternatives I can think of) would be:

* run old/new code on the same machine for each test run (I think OPA does this)
* Running old/new code on different machines, with the both run several times to get some measure of per-machine variance (an attempt to quantify the likelihood of processor skew)
* Assurances that hardware/software stack are identical across runs (might be impossible to guarantee this, since upgrades are likely/expected, and so there will always be the possibility that maintenance has been performed between runs)

I might be missing some alternatives?

Or are we not looking to make conclusions by doing comparative analysis? If not, what are we hoping to use this data for?

Reply:
We are not branching on minor version releases, for example, I do not see a branch named `release-3.9.2`. So, this means we will not be branching on `3.10.1` as well. But let's say we release `3.10.1` in future, then do we update the branch `release-3.10` with the patches of `3.10.1`? Just trying to determine what do we refer to when we say, "old code" and run it on the same machine as the "new code" to compare performance.

Reply:
It's probably good enough to run these only against pull requests to main, since patches are most likely cherrypicks, so at that point the code's performance impact would already be understood (although it is possible for there to be unknown interactions with older code, example: when `--constraint-violations-limit` scaled poorly, that could have compounded some other performance penalty).

Strictly speaking, if you wanted to test every PR, you'd benchmark the PR-as-written against the head of the branch to which it will be applied.

Testing only off of main would avoid the need to build code twice per PR (since we make a build for every commit to main, IIRC).

Reply:
( chiming in to throw my 2cents ) I think benchmarking against the head of the branch would be more incrementally useful as we iterate over a change. 

i.e. "does my latest commit change the performance in this branch" vs "does my latest commit OR something that was checked into `main` change the performance in this branch" ?

Reply:
> Strictly speaking, if you wanted to test every PR, you'd benchmark the PR-as-written against the head of the branch to which it will be applied.

I agree, this is a good idea and is much more accurate than benchmarking everything against head

Reply:
One drawback is though we are running two benchmarking process that increases the time to complete workflow by 10 mins avg.

Reply:
tests seem to be taking 20mins+ now which makes e2e take over twice as long

@acpana should we decrease count to 5? anything else we can do? 

perhaps this can be a manual command (like `/benchmark`)?

Reply:
+1, we could hide the benchmark job behind a manual command

Reply:
Definitely running benchmarks less frequently (or manually) could be a good solution here. It will at least let us know if something went wrong.

Reply:
Done. After this PR is merged, we could run benchmarks on PR with commenting `/benchmark`